flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/SlotSharingExecutionSlotAllocatorFactory.java

```diff
@@ -40,13 +41,16 @@ public SlotSharingExecutionSlotAllocatorFactory(
             PhysicalSlotProvider slotProvider,
             boolean slotWillBeOccupiedIndefinitely,
             PhysicalSlotRequestBulkChecker bulkChecker,
-            Duration allocationTimeout) {
+            Duration allocationTimeout,
+            TaskManagerOptions.TaskManagerLoadBalanceMode taskManagerLoadBalanceMode) {
         this(
                 slotProvider,
                 slotWillBeOccupiedIndefinitely,
                 bulkChecker,
                 allocationTimeout,
-                new LocalInputPreferredSlotSharingStrategy.Factory());
+                taskManagerLoadBalanceMode == TaskManagerOptions.TaskManagerLoadBalanceMode.TASKS
+                        ? new TaskBalancedPreferredSlotSharingStrategy.Factory()
```

Comment:
I see that specifying TASKS means this factory will be used to load balance. I do not see the works task number in the factory implementation. It seems to be trying to colocate tasks in a slot. I am not sure when this would be preferable or not.

Reply:
In fact, the phenomenon of uneven task scheduling results is mainly caused by two stages:
1. tasks->slot
2. slots->taskmanager
Because prior to the initial feature freeze, we anticipated that the entire FLIP would be at risk. Our original intention in initiating this PR was to release this change as early as possible to address the imbalance issue in the task ->slot dimension. But now there is no time risk with this feature, so we plan to release both dimensions of functionality together.
BTW, the optimizations and changes mentioned above are highly likely not to be fixed in the current PR as it may be abandoned.
But we will place the corresponding optimization in the appropriate PR position in the future. If you are willing to help with the review, I would greatly appreciate it! :)